### ISSUE TITLE
Non-WSL windows support

### ISSUE THREAD

**Issue by Schmavery** [freebroccolo/ocaml-language-server#84]
@bsansouci and I have been discussing how to get a user-friendly native ocaml workflow working on windows without depending on WSL. Currently this is the plan:

1) Get all relevant dependencies running on windows. We currently ship these all as part of bsb-native, and they install and work in cmd.exe with 1 `npm install` command:
- [x] ocamlc/ocamlopt etc (along with associated mingw dependencies, flexdll etc)
- [x] refmt
- [x] bsc/bsb
- [ ] merlin

2) Have ocaml-language-server equipped to consume binaries shipped with bs-platform when applicable. This would be basically the same as https://github.com/freebroccolo/ocaml-language-server/issues/52, but adding support for also pulling merlin from the bsb-native installation if it's there.
I could imagine adding a flag to the invocation of ocaml-language-server that tells it to try and look in the *.re file's parents for a `bsconfig.json` or `node_modules` and then look in `node_modules/bs-platform/...` for relevant binaries.

This work is on our roadmap for improving windows support, but I thought I'd create an issue in case anyone wanted to help or give suggestions.

**Comment by freebroccolo** [comment]
I agree that native Windows support would be nice. If the tooling is easy enough to install and configure on native then I think we can get the server to work with that.

The main thing to figure out is what you mention in item (2). Specifically, we need to agree on how to detect native tooling and in what fashion that should be configurable.

There are now a number of different scenarios for how the tooling can be installed that we want to be able to support (described in the table below). Whatever we do for Windows native with bsb-native, we want to ensure that it remains compatibile with the others.

| host   | platform | provenance    | scope           | supported |
|--------|----------|---------------|-----------------|-----------|
| native | all      | unspecified   | server env PATH | yes       |
| native | all      | bsb(-native)  | per project     | not yet   |
| native | all      | esy           | per project     | yes?      |
| native | all      | opam 1.0-2.0  | per user        | not yet   |
| native | all      | opam 2.0      | per project     | not yet   |
| wsl    | windows  | unspecified   | server env PATH | not yet   |
| wsl    | windows  | bsb(-native)  | per project     | not yet   |
| wsl    | windows  | esy           | per project     | not yet   |
| wsl    | windows  | opam 1.0-2.0  | per user        | not yet   |
| wsl    | windows  | opam 2.0      | per project     | not yet   |

**Comment by bsansouci** [comment]
Wow thanks for putting this table together. 

Dumb questions:
- how does the Esy support work? Any way bsb-native can leverage what's done here
- does opam 1/2 support different from "server env in PATH" ? I was under the impression that if merlin was installed through opam it was available globally

What binaries does ocaml-language-server need to know about? 

Would it be possible to either
1) make it look inside `node_modules/bs-platform/lib` for a `merlin.exe`
2) make it take a new flag like `-merlin-path ./node_modules/bs-platform/lib` which would be supplied by the Editor plugin.

**Comment by freebroccolo** [comment]
> how does the Esy support work? Any way bsb-native can leverage what's done here

Have to ask @jordwalke about that one.

> does opam 1/2 support different from "server env in PATH" ? I was under the impression that if merlin was installed through opam it was available globally

The situation with opam is a little complicated. If you have the opam environment initialized in your shell via `.bashrc`, `.profile`, `.zshrc`, etc., then it behaves as in the "server env PATH" case.

The problem is that this is not very reliable. There are differences in when exactly these shell scripts are run that differ between platforms and also whether or not they are used for interactive and/or login shells. The situation becomes even messier when there are additional layers of abstraction involved, like `nvm`. I'd say this problem is responsible for probably 50-75% of the reports I see for when people can't get the server to work with merlin even though they seemingly have all the necessary parts installed.

However, there is an alternative way to run binaries which are installed via opam using the `config exec` command which does not suffer from this problem. For instance, you can start Merlin with `opam config exec ocamlmerlin` and this will work even if the scripts under `~/.opam/opam-init/…` have never been run in the current shell.

The latter approach is obviously more robust and is what I am moving the server toward but it is not implemented just yet.

> What binaries does ocaml-language-server need to know about?

Currently it is just the following:

* `bsb`
* `env`
* `esy`
* `ocamlmerlin`
* `ocp-indent`
* `refmt`

Eventually the list will probably also include the following:

* `ocamlformat`
* `ocp-grep`
* `ocp-index`

> Would it be possible to either
> 
>     make it look inside `node_modules/bs-platform/lib` for a `merlin.exe`
>     make it take a new flag like `-merlin-path ./node_modules/bs-platform/lib` which would be supplied by the Editor plugin.

Well, both of those are possible. I'm not sure that having the flag would be very useful though because the intention is that the server should be started by the editor. So the editor would need to know when to supply this flag. I mean, I suppose that could be tied to a configuration option. I would kind of prefer to have the detection work automatically though.

I think it would be reasonable to have the server look for `bsconfig.json` in the project root and if that is found to then try and use tools located under `node_modules`.

The main issue with making the detection automatic is that it just needs to be clear somehow which tooling configuration the server is actively using. This ties in a little bit with the work @jordwalke did recently but I also want to include a feature that will show a "current configuration" page that lists all of this information explicitly for the active session.

**Comment by freebroccolo** [comment]
Status update: I've started working on getting support for this implemented. It may not make it into the next official release but I will probably make a beta version of the server and vscode extension available so that we can experiment with it and see how it works out.
